proposals/partial-type-inference.md

```diff
@@ -0,0 +1,743 @@
+Although it can be done, the computation can take exponential time which we don't want.
+So it has to be restricted to cases, where the problem can be solved effectively but it still has practical usage.
+- **IDE** - Improvement of the type inference can complicate IDE hints during coding. 
+We should give the user clear and not overwhelming errors when there will be an error and try to provide info that helps him to fix it.
+- **Extensions** - We don't want to make this change blocker for another potential feature in the future. 
+So will want to look ahead to other potential directions, which can be done after this feature.
+
+## Detailed design
+[design]: #detailed-design
+
+### Grammar
+
+> Specification: Original section changed in the following way
+
+The following changes are made in [tokens](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#64-tokens) located in the [grammar](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#62-grammars) section.
+
+> [Identifiers](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#643-identifiers)
+
+- The semantics of an identifier named `_` depends on the context in which it appears:
+  - It can denote a named program element, such as a variable, class, or method, or
+  - It can denote a discard ([§9.2.9.1](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/variables.md#9291-discards)).
+  - **It can denote an inferred type argument avoiding specifying type arguments which can be inferred by the compiler.**
+
+> [Keywords](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#644-keywords)
+
+* A ***contextual keyword*** is an identifier-like sequence of characters that has special meaning in certain contexts, but is not reserved, and can be used as an identifier outside of those contexts as well as when prefaced by the `@` character.
+
+  ```diff
+  contextual_keyword
+      : 'add'    | 'alias'      | 'ascending' | 'async'     | 'await'
+      | 'by'     | 'descending' | 'dynamic'   | 'equals'    | 'from'
+      | 'get'    | 'global'     | 'group'     | 'into'      | 'join'
+      | 'let'    | 'nameof'     | 'on'        | 'orderby'   | 'partial'
+      | 'remove' | 'select'     | 'set'       | 'unmanaged' | 'value'
+  +   | 'var'    | 'when'       | 'where'     | 'yield'     | '_'
+  -   | 'var'    | 'when'       | 'where'     | 'yield'
+    ;
+  ```
+
+### Type arguments
+
+> Specification: Original section changed in the following way
+
+* We change the meaning of the content of *type_argument_list* in two contexts.
+  * [Constructed types](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/types.md#84-constructed-types) occuring in [*object_creation_expression*](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#128162-object-creation-expressions)
```

Comment:
As a general comment, I appreciate the references back to the standard/spec. The format you adopted in other sections (copy a paragraph from original spec and add some highlighted edits) is really nice. But when I look at "Type arguments" section, it's not clear which section of the spec is being edited. 

Maybe you mean to edit section [8.4.2](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/types.md#842-type-arguments) ("type arguments")? That seems to be the right place to introduce ***inferred_type_argument*** into the ***type_argument_list*** grammar.
(If so, then the references to "constructed types" (8.4) and  "object_creation_expression" could be removed)

Below is a possible grammar change to introduce `inferred_type_argument`:
  ```antlr
type_argument_list
    : '<' type_arguments '>'
    ;
    
type_arguments // modified
    : type_argument (',' type_arguments)
    | inferred_type_argument (',' type_arguments)
    ;   

type_argument
    : type
    ;

inferred_type_argument // new
    : '_'
    ;
  ```
Then we can say that ***inferred_type_argument*** is only allowed in the contexts that you listed.

We still need to resolve the issue of back compat:
```
public class _ { }
public class C<T> 
{
    public void M(C<_> x) { } // allowed today
}
```
We'll need to decide whether the presence of an existing type named "_"  in scope affects the interpretation of "_" in type argument lists, or whether "_" always means "inferred type argument" (ie. we take a compat break, maybe warn on types named "_").

Reply:
Should be fixed by [6039b9f](https://github.com/dotnet/csharplang/pull/7582/commits/6039b9fcc71f26a254f65ce4261200d04ba90e17).